bin/reth/src/debug_cmd/count_opcodes.rs

```diff
@@ -0,0 +1,406 @@
+    };
+}
+
+/// `reth count-opcodes` command
+#[derive(Debug, Parser)]
+pub struct Command {
+    /// The database path.
+    #[arg(long, value_name = "DB_DIR", verbatim_doc_comment)]
+    db_dir: PathBuf,
+    /// The top `take` opcodes to display.
+    #[arg(long, value_name = "TAKE_SIZE", verbatim_doc_comment, default_value = "10")]
+    take: usize,
+}
+
+impl Command {
+    /// Execute `count-opcodes` command
+    pub async fn execute(&self) -> eyre::Result<()> {
+        // read db
+        let db = Arc::new(open_db_read_only(self.db_dir.as_path(), None)?);
+        // create spec
+        let spec = Arc::new(ChainSpecBuilder::mainnet().build());
+        // create db provider
+        let factory = ProviderFactory::new(db.clone(), spec.clone());
+        let provider = factory.provider()?;
+
+        // get bytecodes table
+        let bytecodes = provider.table::<tables::Bytecodes>()?;
+
+        let mut opcode_counter = OpCodeCounter::new();
+
+        for (_address, bytecode) in bytecodes {
+            let filtered_bytes = filter_bytecode_bytes(bytecode.bytes());
+            opcode_counter.count_sequences(&filtered_bytes);
+        }
+
+        opcode_counter.print_counts(self.take);
+
+        Ok(())
+    }
+}
+
+/// Takes bytecode bytes and returns filtered bytes without `PUSH` data
+fn filter_bytecode_bytes(bytes: &Bytes) -> Bytes {
+    let mut push_data_to_skip = 0_usize;
+
+    let iter = bytes.iter().filter(|op| {
+        if push_data_to_skip > 0 {
+            push_data_to_skip -= 1;
+            return false;
+        };
+        if (opcode::PUSH1..=opcode::PUSH32).contains(op) {
+            push_data_to_skip = (**op - opcode::PUSH1 + 1) as usize;
+        };
+        true
+    });
+
+    Bytes::from_iter(iter)
+}
+
+struct OpCodeCounter {
+    opcodes: HashMap<[u8; 1], usize>,
+    couples_opcodes: HashMap<[u8; 2], usize>,
+    triplets_opcodes: HashMap<[u8; 3], usize>,
+    quadruplets_opcodes: HashMap<[u8; 4], usize>,
+}
+
+impl OpCodeCounter {
+    fn new() -> Self {
+        Self {
+            opcodes: HashMap::new(),
+            couples_opcodes: HashMap::new(),
+            triplets_opcodes: HashMap::new(),
+            quadruplets_opcodes: HashMap::new(),
+        }
+    }
+
+    fn count_sequences(&mut self, bytes: &Bytes) {
+        for (i, opcode) in bytes.iter().enumerate() {
+            let couple = bytes.get(i..=i + 1);
+            let triple = bytes.get(i..=i + 2);
+            let quadruple = bytes.get(i..=i + 3);
+
+            *self.opcodes.entry([*opcode]).or_default() += 1;
+            if let Some(t) = couple {
+                *self.couples_opcodes.entry([t[0], t[1]]).or_default() += 1;
+            }
+            if let Some(t) = triple {
+                *self.triplets_opcodes.entry([t[0], t[1], t[2]]).or_default() += 1;
+            }
+            if let Some(t) = quadruple {
+                *self.quadruplets_opcodes.entry([t[0], t[1], t[2], t[3]]).or_default() += 1;
+            }
+        }
```

Comment:
nit: can use slice patterns, unfortunately it evals to a slice not a reference to an array so you still have to construct the array for the key

```rust
if let Some(&[a, b]) = bytes.get(i..i+2) {
    *self.couples.entry([a, b]).or_default() += 1;
}
```